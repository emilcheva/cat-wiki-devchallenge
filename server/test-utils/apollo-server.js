const { createTestClient } = require("apollo-server-testing");
const catAPI = require("../datasources/cat-api");
const { ApolloServer } = require("apollo-server");
const resolvers = require('../resolvers');
const typeDefs = require('../schema');

const createApolloTestServer = () => {
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => {
      return {
        catAPI: new catAPI(),
      };
    },
  });
  return server;
};

const apolloServer = createApolloTestServer();

export const { query } = createTestClient(apolloServer);
