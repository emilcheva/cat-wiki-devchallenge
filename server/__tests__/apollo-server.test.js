import nock from "nock";
import { gql } from "apollo-server";
import { query } from "../test-utils/apollo-server";
import breed from "../mocks/breed.json";

describe("Test Cat Wiki API", () => {
  const baseURL = "https://api.thecatapi.com/v1";

  const GET_BREEDS = gql`
    query getBreeds($limit: Int) {
      getBreeds(limit: $limit) {
        name
        id
      }
    }
  `;

  const GET_BREEDS_BY_NAME = gql`
    query getBreedsByName($breedName: String!) {
      getBreedsByName(breedName: $breedName) {
        description
        name
        temperament
        strangerFriendly
        socialNeeds
        healthIssues
        intelligence
        grooming
        childFriendly
        affectionLevel
        adaptability
        lifeSpan
        origin
        id
      }
    }
  `;

  it("should return single breed", async () => {
    nock(baseURL).get("/breeds").query({ limit: 1 }).reply(200, breed);

    const response = await query({
      query: GET_BREEDS,
      variables: { limit: 1 },
    });

    expect(response.errors).toBeUndefined();
    expect(response).toMatchSnapshot();
  });

  it("should return breeds by name", async () => {
    nock(baseURL)
      .get("/breeds/search")
      .query({ q: "persian" })
      .reply(200, breed);

    const response = await query({
      query: GET_BREEDS_BY_NAME,
      variables: { breedName: "persian" },
    });
    expect(response).toMatchSnapshot();
  });

  it("should throw error if no breed name is given", async () => {
    nock(baseURL)
      .get("/breeds/search")
      .query({ q: "persian" })
      .reply(200);

    const response = await query({
      query: GET_BREEDS_BY_NAME
    });

    expect(response.data).toBeUndefined();
  });
});
