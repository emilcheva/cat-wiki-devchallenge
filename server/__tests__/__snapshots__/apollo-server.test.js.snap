// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Cat Wiki API should return breeds by name 1`] = `
Object {
  "data": Object {
    "getBreedsByName": Array [
      Object {
        "adaptability": 5,
        "affectionLevel": null,
        "childFriendly": null,
        "description": "Persians are sweet, gentle cats that can be playful or quiet and laid-back. Great with families and children, they absolutely love to lounge around the house. While they don’t mind a full house or active kids, they’ll usually hide when they need some alone time.",
        "grooming": 5,
        "healthIssues": null,
        "id": "pers",
        "intelligence": 3,
        "lifeSpan": null,
        "name": "Persian",
        "origin": "Iran (Persia)",
        "socialNeeds": null,
        "strangerFriendly": null,
        "temperament": "Affectionate, loyal, Sedate, Quiet",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test Cat Wiki API should return single breed 1`] = `
Object {
  "data": Object {
    "getBreeds": Array [
      Object {
        "id": "pers",
        "name": "Persian",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
